### ISSUE TITLE
Getting RexRay's DO support working.

### ISSUE THREAD

**Issue by Routhinator** [codedellemc/rexray#851]
# Summary
I have just gotten into rexray after using both GlusterFS and NFS with Docker volumes.

I got RexRay setup with Digital Ocean (DOBS) and it's running. So based on the readme the following should work:
```
sudo docker run -ti --volume-driver=rexray -v test:/test busybox
```
But I get 
```
docker: Error response from daemon: create test: Error looking up volume plugin rexray: legacy plugin: plugin not found.
```
There's no containerized plugin for `rexray/dobs` so I'm at a loss on how to get this working.

## Version
```shell
REX-Ray
-------
Binary: /usr/bin/rexray
Flavor: client+agent+controller
SemVer: 0.9.0
OsArch: Linux-x86_64
Branch: v0.9.0
Commit: 2a7458dd90a79c673463e14094377baf9fc8695e
Formed: Wed, 03 May 2017 21:38:11 UTC

libStorage
----------
SemVer: 0.6.0
OsArch: Linux-x86_64
Branch: v0.9.0
Commit: fa055d6da595602715bdfd5541b4aa6d4dcbcbd9
Formed: Wed, 03 May 2017 21:36:11 UTC

```

## Expected Behavior
Successful container start with :
```
sudo docker run -ti --volume-driver=rexray -v test:/test busybox
``` 

## Actual Behavior
```
sudo docker run -ti --volume-driver=rexray -v test:/test busybox
docker: Error response from daemon: create test: Error looking up volume plugin rexray: legacy plugin: plugin not found.
```

## Configuration Files
config.yml
```yaml
# {{ ansible_managed }}
rexray:
  logLevel: {{ rexray_log_level }}
  storageDrivers:
  - dobs
  modules:
    default-docker:
      type: docker
      desc: The default docker module.
      host: unix:///run/docker/plugins/rexray.sock
      libstorage:
        service: dobs

libstorage:
  service: dobs
  server:
    services:
      dobs:
        driver: dobs

dobs:
  token:  {{ vault_do_api_key }}
  region: {{ do_region_slug }}
  statusMaxAttempts: 10
  statusInitialDelay: 100ms
  statusTimeout: 2m
```

## Logs
It is very important when filing an issue to include associated logs. There are
two different logs about which to be concerned: the service log (if REX-Ray is
running as a service) and the client log.

/var/log/rexray.log:
```
                          ⌐▄Q▓▄Ç▓▄,▄_
                         Σ▄▓▓▓▓▓▓▓▓▓▓▄π
                       ╒▓▓▌▓▓▓▓▓▓▓▓▓▓▀▓▄▄.
                    ,_▄▀▓▓ ▓▓ ▓▓▓▓▓▓▓▓▓▓▓█
                   │▄▓▓ _▓▓▓▓▓▓▓▓▓┌▓▓▓▓▓█
                  _J┤▓▓▓▓▓▓▓▓▓▓▓▓▓├█▓█▓▀Γ
            ,▄▓▓▓▓▓▓^██▓▓▓▓▓▓▓▓▓▓▓▓▄▀▄▄▓▓Ω▄
            F▌▓▌█ⁿⁿⁿ  ⁿ└▀ⁿ██▓▀▀▀▀▀▀▀▀▀▀▌▓▓▓▌
             'ⁿ_  ,▄▄▄▄▄▄▄▄▄█_▄▄▄▄▄▄▄▄▄ⁿ▀~██
               Γ  ├▓▓▓▓▓█▀ⁿ█▌▓Ω]█▓▓▓▓▓▓ ├▓
               │  ├▓▓▓▓▓▌≡,__▄▓▓▓█▓▓▓▓▓ ╞█~   Y,┐
               ╞  ├▓▓▓▓▓▄▄__^^▓▓▓▌▓▓▓▓▓  ▓   /▓▓▓
                  ├▓▓▓▓▓▓▓▄▄═▄▓▓▓▓▓▓▓▓▓  π ⌐▄▓▓█║n
                _ ├▓▓▓▓▓▓▓▓▓~▓▓▓▓▓▓▓▓▓▓  ▄4▄▓▓▓██
                µ ├▓▓▓▓█▀█▓▓_▓▓███▓▓▓▓▓  ▓▓▓▓▓Ω4
                µ ├▓▀▀L   └ⁿ  ▀   ▀ ▓▓█w ▓▓▓▀ìⁿ
                ⌐ ├_                τ▀▓  Σ⌐└
                ~ ├▓▓  ▄  _     ╒  ┌▄▓▓  Γ
                  ├▓▓▓▌█═┴▓▄╒▀▄_▄▌═¢▓▓▓  ╚
               ⌠  ├▓▓▓▓▓ⁿ▄▓▓▓▓▓▓▓┐▄▓▓▓▓  └
               Ω_.└██▓▀ⁿÇⁿ▀▀▀▀▀▀█≡▀▀▀▀▀   µ
               ⁿ  .▄▄▓▓▓▓▄▄┌ ╖__▓_▄▄▄▄▄*Oⁿ
                 û▌├▓█▓▓▓██ⁿ ¡▓▓▓▓▓▓▓▓█▓╪
                 ╙Ω▀█ ▓██ⁿ    └█▀██▀▓█├█Å
                     ⁿⁿ             ⁿ ⁿ^
:::::::..  .,::::::    .,::      .::::::::..    :::.  .-:.     ::-.
;;;;'';;;; ;;;;''''    ';;;,  .,;; ;;;;'';;;;   ;;';;  ';;.   ;;;;'
 [[[,/[[['  [[cccc       '[[,,[['   [[[,/[[['  ,[[ '[[,  '[[,[[['
 $$$$$$c    $$""""        Y$$$Pcccc $$$$$$c   c$$$cc$$$c   c$$"
 888b "88bo,888oo,__    oP"''"Yo,   888b "88bo,888   888,,8P"'
 MMMM   "W" """"YUMMM,m"       "Mm, MMMM   "W" YMM   ""'mM"

REX-Ray
-------
Binary: /usr/bin/rexray
Flavor: client+agent+controller
SemVer: 0.9.0
OsArch: Linux-x86_64
Branch: v0.9.0
Commit: 2a7458dd90a79c673463e14094377baf9fc8695e
Formed: Wed, 03 May 2017 21:38:11 UTC

libStorage
----------
SemVer: 0.6.0
OsArch: Linux-x86_64
Branch: v0.9.0
Commit: fa055d6da595602715bdfd5541b4aa6d4dcbcbd9
Formed: Wed, 03 May 2017 21:36:11 UTC

time="2017-05-16T03:45:27Z" level=debug msg="created scoped scope" new=libstorage.server parentScopes="rexray," 
time="2017-05-16T03:45:27Z" level=debug msg="created scoped scope" new=libstorage.server.services.dobs parentScopes="libstorage.server,rexray," 
time="2017-05-16T03:45:28Z" level=debug msg="created scoped scope" new=rexray.modules.default-docker parentScopes="rexray," 
time="2017-05-16T03:45:28Z" level=debug msg="created scoped scope" new=libstorage.client parentScopes="rexray.modules.default-docker,rexray,"
```
journalctl -f output:
https://gist.github.com/Routhinator/8654d324d9e486cd6ce9a1694a932ffc

rexray stdout output:
https://gist.github.com/Routhinator/4794805a4bfc252969ef7dc935561a33

**Comment by akutz** [comment]
The name of the driver is "rexray" from Docker's POV. Then again, you could have defined a config module named dobs, but you didn't provide your config or logs as is requested when filing new issues :) 

Try making "--volume-driver=rexray", and if that doesn't work please provide the information that was requested when you filed this issue.

**Comment by Routhinator** [comment]
@akutz I've updated the question/issue - I was using `--volume-driver=rexray` - `dobs` was just a second guess.

**Comment by Routhinator** [comment]
Also worth noting the service is running, and docker was restarted after it was started successfully:

```shell
root@swarm-master-fra1-01:/var/log/rexray# rexray status
● rexray.service - rexray
   Loaded: loaded (/etc/systemd/system/rexray.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2017-05-15 04:16:40 UTC; 12min ago
 Main PID: 17558 (rexray)
    Tasks: 8
   Memory: 5.6M
      CPU: 1.829s
   CGroup: /system.slice/rexray.service
           └─17558 /usr/bin/rexray start -f

May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: Branch: v0.9.0
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: Commit: 2a7458dd90a79c673463e14094377baf9fc8695e
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: Formed: Wed, 03 May 2017 21:38:11 UTC
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: libStorage
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: ----------
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: SemVer: 0.6.0
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: OsArch: Linux-x86_64
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: Branch: v0.9.0
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: Commit: fa055d6da595602715bdfd5541b4aa6d4dcbcbd9
May 15 04:16:40 swarm-master-fra1-01 rexray[17558]: Formed: Wed, 03 May 2017 21:36:11 UTC

````

**Comment by kacole2** [comment]
@Routhinator try simplifying your config file for now. This should be all that's needed

```
libstorage:
  service: dobs
dobs:
  token: 123456
  region: nyc1
```

**Comment by Routhinator** [comment]
@kacole2 Implemented the config as you suggested, restarted rexray, and then restarted docker:

```
root@swarm-master-fra1-01:/etc/rexray# docker run -ti --volume-driver=rexray -v test:/test busybox
docker: Error response from daemon: create test: Error looking up volume plugin rexray: legacy plugin: plugin not found.
```

Do I need to provide any config to docker itself?

**Comment by kacole2** [comment]
@Routhinator have you started the rexray service with `sudo rexray start`?

**Comment by Routhinator** [comment]
@kacole2 - Yes, I thought that was clear. The service is running and was restarted after the configuration changes, as well as docker.

**Comment by Routhinator** [comment]
@kacole2 Service status output:

```
root@swarm-master-fra1-01:/etc/rexray# service rexray status
● rexray.service - rexray
   Loaded: loaded (/etc/systemd/system/rexray.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2017-05-15 14:03:05 UTC; 1h 1min ago
 Main PID: 28867 (rexray)
    Tasks: 9
   Memory: 7.5M
      CPU: 3.481s
   CGroup: /system.slice/rexray.service
           └─28867 /usr/bin/rexray start -f

May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: Branch: v0.9.0
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: Commit: 2a7458dd90a79c673463e14094377baf9fc8695e
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: Formed: Wed, 03 May 2017 21:38:11 UTC
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: libStorage
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: ----------
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: SemVer: 0.6.0
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: OsArch: Linux-x86_64
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: Branch: v0.9.0
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: Commit: fa055d6da595602715bdfd5541b4aa6d4dcbcbd9
May 15 14:03:05 swarm-master-fra1-01 rexray[28867]: Formed: Wed, 03 May 2017 21:36:11 UTC

```

**Comment by kacole2** [comment]
@Routhinator what docker version are you running?

**Comment by Routhinator** [comment]
@kacole2 

```
root@swarm-master-fra1-01:/etc/rexray# docker --version
Docker version 17.05.0-ce, build 89658be
```

**Comment by clintkitson** [comment]
Chris,

Any chance the Docker daemon is running from within a container? How was
Docker installed on the instance?

**Comment by Routhinator** [comment]
@clintkitson  Docker is installed directly on the host - (not sure why docker would ever be installed in a container?)
Docker was installed directly from the official docker repo using apt on a 16.04 host.

**Comment by clintkitson** [comment]
For example, when LinuxKit creates a Docker OS image, it actually runs Docker using runc. Docker looks inside of the `/run` path for the unix sockets of active plugins. Your error typically means that Docker is not able to see the socket that REX-Ray has created. So either REX is running in a container and not using a shared `/run` sub folder or Docker is running in a container and not using a shared `/run` sub folder.

**Comment by Routhinator** [comment]
@clintkitson I see. Nothing in containers here. Both docker and rexray are running directly on the host, as root:

Nothing running in a container currently, from what I can tell.
```shell
root@swarm-master-fra1-01:/etc/rexray# grep -qa container=lxc /proc/1/environ
root@swarm-master-fra1-01:/etc/rexray# 
```

```shell
29457 root       20   0 53808 21924 16600 S  0.7  1.1  0:21.73 ├─ /usr/bin/gitlab-ci-multi-runner run --working-directory /home/gitlab-runner --config /etc/gitlab-runner/config.toml --service gitlab-runner --syslog --user gitlab-runner
29470 root       20   0 53808 21924 16600 S  0.0  1.1  0:03.96 │  ├─ /usr/bin/gitlab-ci-multi-runner run --working-directory /home/gitlab-runner --config /etc/gitlab-runner/config.toml --service gitlab-runner --syslog --user gitlab-runner
29469 root       20   0 53808 21924 16600 S  0.0  1.1  0:03.42 │  ├─ /usr/bin/gitlab-ci-multi-runner run --working-directory /home/gitlab-runner --config /etc/gitlab-runner/config.toml --service gitlab-runner --syslog --user gitlab-runner
29468 root       20   0 53808 21924 16600 S  0.0  1.1  0:04.18 │  ├─ /usr/bin/gitlab-ci-multi-runner run --working-directory /home/gitlab-runner --config /etc/gitlab-runner/config.toml --service gitlab-runner --syslog --user gitlab-runner
29464 root       20   0 53808 21924 16600 S  0.0  1.1  0:03.94 │  ├─ /usr/bin/gitlab-ci-multi-runner run --working-directory /home/gitlab-runner --config /etc/gitlab-runner/config.toml --service gitlab-runner --syslog --user gitlab-runner
29461 root       20   0 53808 21924 16600 S  0.0  1.1  0:03.92 │  ├─ /usr/bin/gitlab-ci-multi-runner run --working-directory /home/gitlab-runner --config /etc/gitlab-runner/config.toml --service gitlab-runner --syslog --user gitlab-runner
29460 root       20   0 53808 21924 16600 S  0.0  1.1  0:00.00 │  ├─ /usr/bin/gitlab-ci-multi-runner run --working-directory /home/gitlab-runner --config /etc/gitlab-runner/config.toml --service gitlab-runner --syslog --user gitlab-runner
29459 root       20   0 53808 21924 16600 S  0.0  1.1  0:02.19 │  └─ /usr/bin/gitlab-ci-multi-runner run --working-directory /home/gitlab-runner --config /etc/gitlab-runner/config.toml --service gitlab-runner --syslog --user gitlab-runner
29019 root       20   0  503M 56012 30228 S  0.7  2.7  2:04.33 ├─ /usr/bin/dockerd -H fd://
29416 root       20   0  503M 56012 30228 S  0.0  2.7  0:15.17 │  ├─ /usr/bin/dockerd -H fd:// 
29398 root       20   0  503M 56012 30228 S  0.7  2.7  0:15.42 │  ├─ /usr/bin/dockerd -H fd:// 
29038 root       20   0  503M 56012 30228 S  0.7  2.7  0:11.87 │  ├─ /usr/bin/dockerd -H fd:// 
29037 root       20   0  503M 56012 30228 S  0.0  2.7  0:16.72 │  ├─ /usr/bin/dockerd -H fd:// 
29036 root       20   0  503M 56012 30228 S  0.0  2.7  0:14.63 │  ├─ /usr/bin/dockerd -H fd:// 
29034 root       20   0  503M 56012 30228 S  0.0  2.7  0:15.85 │  ├─ /usr/bin/dockerd -H fd:// 
29027 root       20   0  276M 14852  6988 S  0.0  0.7  0:22.90 │  ├─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/contai
29482 root       20   0  276M 14852  6988 S  0.0  0.7  0:04.07 │  │  ├─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/con
29396 root       20   0  276M 14852  6988 S  0.0  0.7  0:04.75 │  │  ├─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/con
29035 root       20   0  276M 14852  6988 S  0.0  0.7  0:04.34 │  │  ├─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/con
29032 root       20   0  276M 14852  6988 S  0.0  0.7  0:02.67 │  │  ├─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/con
29031 root       20   0  276M 14852  6988 S  0.0  0.7  0:04.66 │  │  ├─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/con
29030 root       20   0  276M 14852  6988 S  0.0  0.7  0:00.00 │  │  ├─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/con
29029 root       20   0  276M 14852  6988 S  0.0  0.7  0:00.00 │  │  ├─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/con
29028 root       20   0  276M 14852  6988 S  0.0  0.7  0:02.33 │  │  └─ docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/con
29026 root       20   0  503M 56012 30228 S  0.0  2.7  0:00.10 │  ├─ /usr/bin/dockerd -H fd:// 
29025 root       20   0  503M 56012 30228 S  0.0  2.7  0:00.00 │  ├─ /usr/bin/dockerd -H fd:// 
29024 root       20   0  503M 56012 30228 S  0.0  2.7  0:00.00 │  ├─ /usr/bin/dockerd -H fd:// 
29023 root       20   0  503M 56012 30228 S  0.0  2.7  0:13.34 │  ├─ /usr/bin/dockerd -H fd:// 
29022 root       20   0  503M 56012 30228 S  0.7  2.7  0:21.12 │  └─ /usr/bin/dockerd -H fd:// 
28867 root       20   0  275M 28844 21412 S  0.0  1.4  0:09.88 ├─ /usr/bin/rexray start -f
28877 root       20   0  275M 28844 21412 S  0.0  1.4  0:00.00 │  ├─ /usr/bin/rexray start -f 
28876 root       20   0  275M 28844 21412 S  0.0  1.4  0:01.47 │  ├─ /usr/bin/rexray start -f 
28875 root       20   0  275M 28844 21412 S  0.0  1.4  0:01.92 │  ├─ /usr/bin/rexray start -f 
28874 root       20   0  275M 28844 21412 S  0.0  1.4  0:00.54 │  ├─ /usr/bin/rexray start -f 
28873 root       20   0  275M 28844 21412 S  0.0  1.4  0:00.00 │  ├─ /usr/bin/rexray start -f 
28871 root       20   0  275M 28844 21412 S  0.0  1.4  0:01.98 │  ├─ /usr/bin/rexray start -f 
28870 root       20   0  275M 28844 21412 S  0.0  1.4  0:00.00 │  ├─ /usr/bin/rexray start -f 
28869 root       20   0  275M 28844 21412 S  0.0  1.4  0:01.75 │  └─ /usr/bin/rexray start -f 

```

**Comment by Routhinator** [comment]
I was really hoping this was user error, but it feels like we are moving towards possible bug. If anyone wants to peek at the live instance to see if they can spot something I'm not thinking of, they are welcome. These are throwaway testing boxes being built with ansible, I can recreate them after.

**Comment by Routhinator** [comment]
Found the cause of the issue, but not the reason yet.

RexRay is not opening a unix socket:
```shell
root@swarm-master-fra1-01:/etc/rexray# ss -l | grep rexray
root@swarm-master-fra1-01:/etc/rexray# 
```

I tried reverting to my more complex config from the simple one @kacole2 suggested and restarted, however I'm still not able to grep a socket for it, despite a successful service start. 

Digging through docs to see ways to define a socket to see if any of the methods resolve.

**Comment by akutz** [comment]
Hi @Routhinator,

Could you please update this issue with the information I [requested above](https://github.com/codedellemc/rexray/issues/851#issuecomment-301345819) (and that is requested when a new issue is filed)?

1. Please enable *debug* logging.
2. Please copy the full log file into a gist.
3. Please paste a link to the gist in a reply to this comment.

I am glad @kacole2 and @clintkitson spent time trying to resolve this issue with you, but what should have occurred beforehand was you providing the requested logs. Here is the template that is pasted into a New Issue for your convenience:

> # Summary
> Please enter a summary of the issue here.
> 
> # New Feature
> This section is for issues that relate to suggested enhancements or other
> ideas that may improve REX-Ray. Please provide as much detail as possible
> regarding the idea. This issue will then serve as the means to have a
> discussion about your idea!
> 
> # Bug Reports
> This section is for issues that relate to discovered problems or bugs.
> 
> ## Version
> Please paste the output of `rexray version`. For example:
> 
> ```shell
> $ rexray version
> REX-Ray
> -------
> Binary: /usr/bin/rexray
> Flavor: client+agent+controller
> SemVer: 0.7.0
> OsArch: Linux-x86_64
> Branch: v0.7.0
> Commit: a20a838ca70838a914b632637398824fcb10d0db
> Formed: Mon, 23 Jan 2017 10:14:32 EST
> 
> libStorage
> ----------
> SemVer: 0.4.0
> OsArch: Linux-x86_64
> Branch: v0.7.0
> Commit: a1103d3f215117f7b9f51dae2b24f852c9c54995
> Formed: Mon, 23 Jan 2017 10:14:12 EST
> ```
> 
> ## Expected Behavior
> Please describe in detail the expected behavior.
> 
> ## Actual Behavior
> Please describe in detail the actual behavior.
> 
> ## Steps To Reproduce
> Please list the steps to reproduce the issue in this section.
> 
> 1. The first step should always be enabling `debug` logging.
>   * Open the file `/etc/rexray/config.yml`
>   * Set the log level for both REX-Ray and libStorage to `debug` and enable
> HTTP request and response tracing for libStorage:
> ```yaml
> rexray:
>   logLevel:        debug
> libstorage:
>   logging:
>     level:         debug
>     httpRequests:  true
>     httpResponses: true
> ```
> * Please list each step with as much detail as possible.
> * The more information gathered up front, the easier it is to solve
> the problem.
> * Thank you!
> 
> ## Configuration Files
> Please paste any related configuration files, such as `/etc/rexray/config.yml`
> in this section. Please use the appropriate formatting when pasting YAML.
> content. For example:
> 
> ```yaml
> rexray:
>   logLevel:        debug
> libstorage:
>   logging:
>     level:         debug
>     httpRequests:  true
>     httpResponses: true
>   service:         ebs
> ebs:
>   accessKey:       123456
>   secretKey:       abcdef
> ```
> 
> Proper formatting of pasted content is very important as structured data can
> sometimes be accidentally recorded incorrectly, affecting the desired outcome.
> 
> ## Logs
> It is very important when filing an issue to include associated logs. There are
> two different logs about which to be concerned: the service log (if REX-Ray is
> running as a service) and the client log.
> 
> ### Service Log
> The REX-Ray service log file is stored at `/var/log/rexray/rexray.log`. Instead
> of pasting the entire log file into this issue, please create a new
> [gist](https://gist.github.com/) and paste the log file's contents there.
> Please name the file `rexray-service.log` in the gist. The proper extension will
> indicate how to format the contents.
> 
> ### Client Log
> The REX-Ray client will emit all of its logs to the console when operating with
> debug logging enabled. Simply copy the contents of the console and paste them
> into the same gist as above naming the file `rexray-client.log.sh`. The `sh`
> extension will cause the contents to be formatted as if they were emitted
> to the shell, which they were.
>

**Comment by akutz** [comment]
Hi @Routhinator,

I understand you pasted information above as a response to my request, but it wasn't a debug log. It was an empty log. I'm not sure how you're deploying RR or where the log is in your case, but the log you provided lacked any content beyond the startup message. 

I'm not sure what or where the disconnect is in what is being configured and then logged, but that's not it :(

**Comment by Routhinator** [comment]
@akutz Sorry I hadn't replied as I was getting more info. I got the `journalctl -f` output of the restart. What I put into the log is acutally the `debug` version of the `/var/log/rexray.log` file, but it seems `journalctl` captures all the important info.

**Comment by Routhinator** [comment]
Also as I had discovered before collecting this info (sorry about not adding it sooner, I somehow missed the debug step in my after-work rush to update this yesterday) it is creating a socket... but I cannot grep it the documentation's grep as it seems to be a randomly serialized libStorage socket. I don't see the docker rexray socket that is expected. Seems it's only exposing libStorage service and not the docker one.

**Comment by akutz** [comment]
Hi @Routhinator,

Hmm, that's not good since the output from that doesn't seem to have any more information either. Is it possible for you to try running the service in the foreground? Something is being lost in translation here between your config and the output. 

`$ rexray service start -f -l debug`

The above command will pick up any global config, but it will also run the service in the foreground, logging to STDOUT (and STDERR where appropriate).

**Comment by akutz** [comment]
Hi @Routhinator,

There are two sockets in play here. The libStorage socket is not related to the RR one, rather the libStorage socket is created for libStorage clients to use. RR embeds both the LS client and server and will launch an embedded version of the latter if not configured to use a remote one.

**Comment by Routhinator** [comment]
@akutz Yes I had learned that and suspected as much. I have a forground GIST here, updating issue with a link as well. https://gist.github.com/Routhinator/4794805a4bfc252969ef7dc935561a33 

Still wathing to see if more comes out.. seems it's waiting for something indefinitely.

**Comment by akutz** [comment]
Hi @Routhinator,

Yes, it looks like it's waiting on a [lock file](https://gist.github.com/Routhinator/4794805a4bfc252969ef7dc935561a33#file-rexray-stdout-log-L178).

Stop the service and then `$ rm /var/lib/libstorage/run/lsx.lock`

After a bunch of false starts it can get stuck.

**Comment by akutz** [comment]
Hi @Routhinator,

Oh wait, maybe it's not stuck. Maybe it's executing a DOBS status check after that entry and you've configured DOBS for 10 attempts at a timeout of 2min each. Wouldn't that take up to 20 mins before it quit trying to contact the DOBS API?

**Comment by Routhinator** [comment]
Good thought, however I removed that, so it's using the defaults. Back to simple config now, sorry for the confusion. 

Also that lock doesn't exist:

```
root@swarm-master-fra1-01:/var/log/rexray#  rm /var/lib/libstorage/run/lsx.lock
rm: cannot remove '/var/lib/libstorage/run/lsx.lock': No such file or directory
```

Looking in the /run/libstorage dir now.

**Comment by Routhinator** [comment]
Aha - there was a `/run/libstorage/lsx.lock`

Removed it and trying again.

**Comment by Routhinator** [comment]
I have a socket - trying the container but betting it will work. 

```
root@swarm-master-fra1-01:/var/log/rexray# ss -l | grep rexray
u_str  LISTEN     0      128    /run/docker/plugins/rexray.sock 314564                * 0 
```

All this over a lock file... :(  Seems like the service should have a failure built in for a stale lock.

**Comment by akutz** [comment]
Ah, yeah, I gave you the wrong path. It's `/var/run/libstorage/lsx.lock` and on some distros `/var/run` is a symlink to `/run`.

**Comment by Routhinator** [comment]
Yep, all is working. Was just the lockfile. Lesson learned. Thank you for all the help. 

I would make this a feature request for a lock file failure, however I suspect it is libStorage, and not RexRay, that requires this.

**Comment by akutz** [comment]
Hi @Routhinator,

Well, the thing is...what's a stale lock? A lock file exists for a good reason, and yes, when it goes bad it's annoying, but it happens. Not often, and only when a series of failed start attempts occur in quick succession. Also, if the logging was working correctly (not sure what is going on as it certainly does work fine over here with SystemD and as well in all of our tests) you would have seen this *much* earlier.

The lock file is cleaned up pretty consistently. As I said, it's an edge case we rarely see unless attempting to cause it on purpose.

**Comment by Routhinator** [comment]
Perhaps I worded that incorrectly. Most services will fail to start if they detect their lock file already exists, whereas rexray will still start, but wait for the lock indefinitely. Seems an odd behaviour, as an existing lock file should indicate the service is running, or died unexpectedly in some cases, and starting is not usually desirable in that case.

**Comment by akutz** [comment]
Hi @Routhinator,

That's a fair point, however there is a reason for it. It's not a service lock file. It's a lock file for the executor -- a binary the LS client downloads from the LS server. The lock file is to ensure that multiple LS clients don't attempt to download the binary at once or update it while it's being checked to see if an update is necessary.

Now maybe you see...it's used by all libStorage clients that can write to its location (non-root users can relocate the libStorage data dirs).

**Comment by Routhinator** [comment]
Got it, that clears it up. Thank you once again.
